tiledb/sm/cpp_api/query.h

```diff
@@ -2316,6 +2316,8 @@ class Query {
         return "READ";
       case TILEDB_WRITE:
         return "WRITE";
+      case TILEDB_WRITE_EXCLUSIVE:
```

Comment:
I'm not convinced we need a new open mode. Why do we need to open the array? If we are using the vacuum functionality under the hood, should we treat this close to the vacuum APIs which do not work on an open array?

Reply:
Vacuuming works on an open array; _consolidation_ does not. A user may close the array to consolidate before reopening to call `delete_fragments` if they so choose. 

The new `QueryType` ensures that delete_fragments be the only process with write access to the array. With `QueryType::WRITE`, the same address may be registered to the multimap more than once. With `QueryType::WRITE_EXCLUSIVE`, we ensure that only _one_ array be registered at a unique address at a time. This will ensure that "everything else" is closed, and delete fragments is the only thing "open" on the array.

@eric-hughes-tiledb or @ihnorton please feel free to correct me or clarify anything I may have missed

Reply:
Every operation on the database, both reading and writing, should be performed on an open array. This is a bare minimum to allow anything like a consistency system to operate. It doesn't matter whether the operation is a read, a write, or rearrangement of data representation (like consolidation and vacuuming). Unless you've got everything tracked by opening it, you can't catch possible consistency conflicts.

It's *desirable* that everything that's opened behind the scenes be explicitly opened by the user. To do otherwise violates the rule of least surprise with a (future) user finding out about consistency conflicts coming out of the blue. To the extent our current API doesn't do this, it should be changed.

It's also *desirable* that every operation that requires exclusive access behind the scenes be opened by the user for exclusive access, and for the same reason for requiring explicit opens. Perhaps in the future we might have a more sophisticated consistency checker that is able to partition the database so that fragment deletes don't require exclusive access. If that happens, that's when we can allow opening the array without requiring exclusivity.

Reply:
Note: change to `QueryType::OPEN_EXCLUSIVE` after discussion.